pkg/scheduler/plugins/deviceshare/utils.go

```diff
@@ -0,0 +1,209 @@
+limitations under the License.
+*/
+
+package deviceshare
+
+import (
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/klog/v2"
+
+	apiext "github.com/koordinator-sh/koordinator/apis/extension"
+	schedulingv1alpha1 "github.com/koordinator-sh/koordinator/apis/scheduling/v1alpha1"
+)
+
+const (
+	nvidiaGpuExist = 1 << iota
+	koordGpuExist
+	gpuCoreExist
+	gpuMemoryExist
+	gpuMemoryRatioExist
+)
+
+var deviceResourceNames = map[schedulingv1alpha1.DeviceType][]corev1.ResourceName{
+	schedulingv1alpha1.GPU:  {apiext.NvidiaGPU, apiext.KoordGPU, apiext.GPUCore, apiext.GPUMemory, apiext.GPUMemoryRatio},
+	schedulingv1alpha1.RDMA: {apiext.KoordRDMA},
+	schedulingv1alpha1.FPGA: {apiext.KoordFPGA},
+}
+
+func hasDeviceResource(podRequest corev1.ResourceList, deviceType schedulingv1alpha1.DeviceType) bool {
+	if podRequest == nil || len(podRequest) == 0 {
+		klog.Warningf("skip checking hasDeviceResource, because pod request is empty")
+		return false
+	}
+	for _, resourceName := range deviceResourceNames[deviceType] {
+		if _, ok := podRequest[resourceName]; ok {
+			return true
+		}
+	}
+	klog.V(5).Infof("pod does not request %v resource", deviceType)
+	return false
+}
+
+func validateCommonDeviceRequest(podRequest corev1.ResourceList, deviceType schedulingv1alpha1.DeviceType) error {
+	if podRequest == nil || len(podRequest) == 0 {
+		return fmt.Errorf("pod request should not be empty")
+	}
+	var commonDevice resource.Quantity
+	switch deviceType {
+	case schedulingv1alpha1.FPGA:
+		commonDevice = podRequest[apiext.KoordFPGA]
+	case schedulingv1alpha1.RDMA:
+		commonDevice = podRequest[apiext.KoordRDMA]
+	default:
+		return fmt.Errorf("device type %v is not supported yet", deviceType)
+	}
+	if commonDevice.Value() > 100 && commonDevice.Value()%100 != 0 {
+		return fmt.Errorf("failed to validate koordinator.sh/%v: %v", deviceType, commonDevice.Value())
+	}
+	return nil
+}
+
+// validateGPURequest uses binary to store each request status.
+// For example, 00010 stands for koordinator.sh/gpu exists, and vice versa.
+// only 00001 || 00010 || 10100 || 01100 are valid GPU request combination
+func validateGPURequest(podRequest corev1.ResourceList) (uint, error) {
+	var gpuCombination uint
+
+	if podRequest == nil || len(podRequest) == 0 {
+		return gpuCombination, fmt.Errorf("pod request should not be empty")
+	}
+
+	if _, exist := podRequest[apiext.NvidiaGPU]; exist {
+		gpuCombination |= nvidiaGpuExist
+	}
+	if _, exist := podRequest[apiext.KoordGPU]; exist {
+		gpuCombination |= koordGpuExist
+	}
+	if gpuCore, exist := podRequest[apiext.GPUCore]; exist {
+		// koordinator.sh/gpu-core should be something like: 25, 50, 75, 100, 200, 300
+		if gpuCore.Value() > 100 && gpuCore.Value()%100 != 0 {
+			return gpuCombination, fmt.Errorf("failed to validate koordinator.sh/gpu: %v", gpuCore)
+		}
+		gpuCombination |= gpuCoreExist
+	}
+	if _, exist := podRequest[apiext.GPUMemory]; exist {
```

Comment:
memoryRatio should also check if > 100; as for gpu-mem, due to depends on concrete nodes,  so  not check is ok for now.